pkg/apis/kappctrl/v1alpha1/types.go

```diff
@@ -34,28 +41,46 @@ type AppList struct {
 
 // +k8s:openapi-gen=true
 type AppSpec struct {
-	ServiceAccountName string      `json:"serviceAccountName,omitempty"`
-	Cluster            *AppCluster `json:"cluster,omitempty"`
+	// +optional
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
 
-	Fetch    []AppFetch    `json:"fetch,omitempty"`
+	// +optional
+	Cluster *AppCluster `json:"cluster,omitempty"`
+
+	// +kubebuilder:validation:MinItems=1
+	Fetch []AppFetch `json:"fetch,omitempty"`
+
+	// +kubebuilder:validation:MinItems=1
 	Template []AppTemplate `json:"template,omitempty"`
-	Deploy   []AppDeploy   `json:"deploy,omitempty"`
+
```

Comment:
Curious what error message something like this gives? I know we have some specific error messages for the case of empty templates/deploys/fetches, so if those are better or more informative, may be we leave off the validations in the schema?

Reply:
Yeah, it's a good call out. The error from this would be worse:

```
App.kappctrl.k14s.io "simple-app" is invalid: spec.deploy:
    Invalid value: "null": spec.deploy in body must be of type array: "null" (reason: Invalid)
```

versus

```
 Expected exactly one deploy option
```

An advantage would be that the App itself is not created if it is invalid, but maybe we should preserve the behavior for now and weigh whether it would be worthwhile to enforce this via schema.

Reply:
Yeah, good call. It actually is a bit nicer, so maybe it would be worthwhile. Personally prefer the error kapp-controller currently gives, but the message below is okay:

```
App.kappctrl.k14s.io "simple-app" is invalid: spec.deploy:
      Invalid value: 0: spec.deploy in body should have at least 1 items (reason: Invalid)
```